ompi/mpiext/affinity/c/mpiext_affinity_str.c

```diff
@@ -94,6 +94,199 @@ int OMPI_Affinity_str(ompi_affinity_fmt_t fmt_type,
+        }
+
+        /* Go upward and find the socket this PU belongs to */
+        socket = pu;
+        while (NULL != socket && socket->type != HWLOC_OBJ_SOCKET) {
+            socket = socket->parent;
+        }
+        socket_index = 0;
+        if (NULL != socket) {
+            socket_index = socket->logical_index;
+        }
+
+        /* Save this socket/core/pu combo.  LAZY: Assuming that we
+           won't have more PU's per core than (sizeof(int)*8). */
+        data[socket_index][core_index] |= (1 << pu->sibling_rank);
+    }
+
+    *map = data;
+    return OMPI_SUCCESS;
+}
+
+/*
+ * Turn an int bitmap to a "a-b,c" range kind of string
+ */
+static char *bitmap2rangestr(int bitmap)
+{
+    size_t i;
+    int range_start, range_end;
+    bool first, isset;
+    char tmp[BUFSIZ];
+    const int stmp = sizeof(tmp) - 1;
+    static char ret[BUFSIZ];
+
+    memset(ret, 0, sizeof(ret));
+
+    first = true;
+    range_start = -999;
+    for (i = 0; i < sizeof(int) * 8; ++i) {
+        isset = (bitmap & (1 << i));
+
+        /* Do we have a running range? */
+        if (range_start >= 0) {
+            if (isset) {
+                continue;
+            } else {
+                /* A range just ended; output it */
+                if (!first) {
+                    strncat(ret, ",", sizeof(ret) - strlen(ret) - 1);
+                } else {
+                    first = false;
+                }
+
+                range_end = i - 1;
+                if (range_start == range_end) {
+                    snprintf(tmp, stmp, "%d", range_start);
+                } else {
+                    snprintf(tmp, stmp, "%d-%d", range_start, range_end);
+                }
+                snprintf(ret + strlen(ret), BUFSIZ, "%s", tmp);
+
+                range_start = -999;
+            }
+        }
+
+        /* No running range */
+        else {
+            if (isset) {
+                range_start = i;
+            }
+        }
+    }
+
+    /* If we ended the bitmap with a range open, output it */
+    if (range_start >= 0) {
+        if (!first) {
+            strncat(ret, ",", sizeof(ret) - strlen(ret) - 1);
+            first = false;
+        }
+
+        range_end = i - 1;
+        if (range_start == range_end) {
+            snprintf(tmp, stmp, "%d", range_start);
+        } else {
+            snprintf(tmp, stmp, "%d-%d", range_start, range_end);
+        }
+        snprintf(ret + strlen(ret), BUFSIZ, "%s",  tmp);
+    }
+
+    return ret;
+}
+
+/*
+ * Make a prettyprint string for a hwloc_cpuset_t
+ */
+static int cset2str(char *str, int len, hwloc_topology_t topo, hwloc_cpuset_t cpuset)
+{
+    bool first;
+    int num_sockets, num_cores;
+    int ret, socket_index, core_index;
+    char tmp[BUFSIZ];
+    const int stmp = sizeof(tmp) - 1;
+    int **map = NULL;
+
+    str[0] = tmp[stmp] = '\0';
+
+    /* if the cpuset is all zero, then not bound */
+    if (hwloc_bitmap_iszero(cpuset)) {
+        return -1;
```

Comment:
Why -1 and not a named enum?

Reply:
I wrestled with what to return here since there was no OMPI_ERRN_NOT_BOUND - I just added it. You're right, better than returning -1.